sdks/java/harness/src/main/java/org/apache/beam/fn/harness/data/QueuingBeamFnDataGrpcClient.java

```diff
@@ -0,0 +1,171 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A {@link BeamFnDataClient} that queues elements so that they can be consumed and processed. In
+ * the thread which calls drainAndBlock.
+ */
+public class QueuingBeamFnDataGrpcClient implements BeamFnDataClient {
+
+  private static final Logger LOG = LoggerFactory.getLogger(QueuingBeamFnDataGrpcClient.class);
+
+  private final BeamFnDataClient mainClient;
+  private final SynchronousQueue<ConsumerAndData> queue;
+  private final HashSet<InboundDataClient> idcs;
+
+  public QueuingBeamFnDataGrpcClient(BeamFnDataClient mainClient) {
+    this.mainClient = mainClient;
+    this.queue = new SynchronousQueue<>();
+    // TODO does this need to be a concurrent hash map (set doesn't seem to exist).
+    this.idcs = new HashSet<InboundDataClient>();
+  }
+
+  /**
+   * Registers the following inbound stream consumer for the provided instruction id and target.
+   *
+   * <p>The provided coder is used to decode elements on the inbound stream. The decoded elements
+   * are passed to the provided consumer. Any failure during decoding or processing of the element
+   * will complete the returned future exceptionally. On successful termination of the stream
+   * (signaled by an empty data block), the returned future is completed successfully.
+   */
+  @Override
+  public <T> InboundDataClient receive(
+      ApiServiceDescriptor apiServiceDescriptor,
+      LogicalEndpoint inputLocation,
+      Coder<WindowedValue<T>> coder,
+      FnDataReceiver<WindowedValue<T>> consumer) {
+    LOG.debug(
+        "Registering consumer for instruction {} and target {}",
+        inputLocation.getInstructionId(),
+        inputLocation.getTarget());
+
+    QueueingFnDataReceiver<T> newConsumer = new QueueingFnDataReceiver<T>(consumer);
+    InboundDataClient idc =
+        this.mainClient.receive(apiServiceDescriptor, inputLocation, coder, newConsumer);
+    newConsumer.idc = idc;
+    this.idcs.add(idc);
+    return idc;
+  }
+
+  private boolean AllDone() {
+    boolean allDone = true;
+    for (InboundDataClient idc : idcs) {
+      allDone &= idc.isDone();
+    }
+    return allDone;
+  }
+
+  /**
+   * Drains the internal queue of this class, by waiting for all WindowValues to be passed to thier
+   * consumers. The thread which wishes to process() the elements should call this method, as this
+   * will cause the consumers to invoke element processing. All receive() and send() calls must be
+   * made prior to calling drainAndBlock, in order to properly terminate.
+   */
+  public void drainAndBlock() throws Exception {
+    // Note: We just throw the exception here
+    // TODO review the error handling here
+    while (true) {
+      ConsumerAndData tuple = null;
+      tuple = queue.poll(50, TimeUnit.MILLISECONDS);
+      // TODO should we implement a timeout logic here? What happens if the
+      // putting thread throws an exception? Can we assume the InboundDataClient will be marked
```

Comment:
Yes, the InboundDataClient should be completed exceptionally.

Reply:
Done